Provide a way to send a JSON object with WebClient

karellen-kim opened the pull request:
Motivation:
It would also be nice to have APIs that automatically convert a request body into bytestring and converts a response into an object.

Modifications:
- Add method `ofJson` in `HttpRequest`
- Add method `aggregateAs` in `HttpResponse`

Result:
- Closes #3707
- Users can convert a request body into bytestring automatically : 
```java
HttpRequest.ofJson(HttpMethod.POST, "/receiveContent", requestContent)
```
- Users can convert a response into an object automatically :
```java
Content content = HttpResponse.of(MediaType.JSON_UTF_8, contentString)
                              .aggregateAs(Content.class)
                              .join(); 

ErrorMessage error = HttpResponse.of(MediaType.JSON_UTF_8, contentString)
                                 .aggregateAs(HttpStatusClass.SERVER_ERROR, ErrorMessage.class)
                                 .join();

ErrorMessage error = HttpResponse.of(MediaType.JSON_UTF_8, contentString)
                                 .aggregateAs(status -> status != HttpStatus.OK, ErrorMessage.class)
                                 .join();

Wrapper<Content> wrapperContent = HttpResponse.of(MediaType.JSON_UTF_8, contentString)
                                              .aggregateAs(new TypeReference<Wrapper<Content>>(){})
                                              .join();
```

karellen-kim committed:
Provide a way to send a JSON object with WebClient (#3707)

karellen-kim committed:
Merge remote-tracking branch 'upstream/master' into json

karellen-kim committed:
Provide a way to send a JSON object with WebClient (#3707)

karellen-kim committed:
Provide a way to send a JSON object with WebClient (#3707)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=3757) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=3757) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=3757) it.</sub>

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+     *
+     * @param valueTypeRef the type reference for decoded response
+     */
+    default <T> CompletableFuture<T> aggregateAs(TypeReference<T> valueTypeRef) {
```
nit: `requireNonNull(valueTypeRef, "valueTypeRef");`

karellen-kim replied:
Thanks!

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+     * @param expectedStatus the expected HTTP status
+     * @param valueTypeRef the type reference for decoded response
+     */
+    default <T> CompletableFuture<T> aggregateAs(HttpStatus expectedStatus, TypeReference<T> valueTypeRef) {
+        return aggregateAs(status -> status == expectedStatus, valueTypeRef, defaultSubscriberExecutor());
```
```suggestion
    default <T> CompletableFuture<T> aggregateAs(HttpStatus expectedStatus, TypeReference<T> valueTypeRef) {
        requireNonNull(expectedStatus, "expectedStatus");
        return aggregateAs(status -> status == expectedStatus, valueTypeRef, defaultSubscriberExecutor());
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+     * @param expectedStatus the expected HTTP status
+     * @param clazz the class for decoded response
+     */
+    default <T> CompletableFuture<T> aggregateAs(HttpStatus expectedStatus, Class<? extends T> clazz) {
+        return aggregateAs(status -> status == expectedStatus, clazz, defaultSubscriberExecutor());
```
```suggestion
    default <T> CompletableFuture<T> aggregateAs(HttpStatus expectedStatus, Class<? extends T> clazz) {
        requireNonNull(expectedStatus, "expectedStatus");
        return aggregateAs(status -> status == expectedStatus, clazz, defaultSubscriberExecutor());
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+     * @param expectedStatusClass the expected class of HTTP status
+     * @param clazz the class for decoded response
+     */
+    default <T> CompletableFuture<T> aggregateAs(HttpStatusClass expectedStatusClass, Class<? extends T> clazz) {
+        return aggregateAs(status -> expectedStatusClass.contains(status), clazz, defaultSubscriberExecutor());
```
```suggestion
    default <T> CompletableFuture<T> aggregateAs(HttpStatusClass expectedStatusClass, Class<? extends T> clazz) {
        requireNonNull(expectedStatusClass, "expectedStatusClass");
        return aggregateAs(status -> expectedStatusClass.contains(status), clazz, defaultSubscriberExecutor());
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+     * @param expectedStatusClass the expected class of HTTP status
+     * @param valueTypeRef the type reference for decoded response
+     */
+    default <T> CompletableFuture<T> aggregateAs(HttpStatusClass expectedStatusClass, TypeReference<T> valueTypeRef) {
+        return aggregateAs(status -> expectedStatusClass.contains(status), valueTypeRef, defaultSubscriberExecutor());
```
```suggestion
    default <T> CompletableFuture<T> aggregateAs(HttpStatusClass expectedStatusClass, TypeReference<T> valueTypeRef) {
        requireNonNull(expectedStatusClass, "expectedStatusClass");
        return aggregateAs(status -> expectedStatusClass.contains(status), valueTypeRef, defaultSubscriberExecutor());
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+     */
+    default <T> CompletableFuture<T> aggregateAs(Predicate<HttpStatus> httpStatusFilter,
+                                                 Class<? extends T> clazz,
+                                                 EventExecutor executor) {
+        return aggregateAs(httpStatusFilter, (content) -> {
```
```suggestion
                                                 EventExecutor executor) {
        requireNonNull(httpStatusFilter, "httpStatusFilter");
        requireNonNull(clazz, "clazz");
        requireNonNull(executor, "executor");
        return aggregateAs(httpStatusFilter, (content) -> {
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+                                                 EventExecutor executor) {
+        return aggregateAs(httpStatusFilter, (content) -> {
+            try {
+                return JacksonUtil.newDefaultObjectMapper().readValue(content, clazz);
```
`JacksonUtil.newDefaultObjectMapper()` creates new object every time. How about adding `readValue()` to `JacksonUtil` as a static method?

karellen-kim replied:
I missed that point. Many thanks!

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+                                                 EventExecutor executor) {
+        return aggregateAs(httpStatusFilter, (content) -> {
+            try {
+                return JacksonUtil.newDefaultObjectMapper().readValue(content, valueTypeRef);
```
Ditto. How about adding readValue() to JacksonUtil as a static method?

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+    default <T> CompletableFuture<T> aggregateAs(Predicate<HttpStatus> httpStatusFilter,
+                                                 TypeReference<T> valueTypeRef,
+                                                 EventExecutor executor) {
+        return aggregateAs(httpStatusFilter, (content) -> {
```
```suggestion
        return aggregateAs(httpStatusFilter, content -> {
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+    default <T> CompletableFuture<T> aggregateAs(Predicate<HttpStatus> httpStatusFilter,
+                                                 Function<String, T> fn,
+                                                 EventExecutor executor) {
+        return aggregate(executor).thenApply(response ->
+        {
```
```suggestion
        return aggregate(executor).thenApply(response -> {
```

ikhoon replied:
Don't we need to check the `Content-Type` of `ResponseHeaders`?
We should convert a response body into JSON if `response.headers().contentType().isJson()` is true.
How about throwing `UnsupportedMediaTypeException` if not matched?

ikhoon commented on core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:
```diff
@@ -603,6 +605,163 @@ static HttpResponse ofFailure(Throwable cause) {
+        }, executor);
+    }
+
+    default <T> CompletableFuture<T> aggregateAs(Predicate<HttpStatus> httpStatusFilter,
+                                                 Function<String, T> fn,
```
HttpStatus may be useful to implement `jsonDecoder`.
```suggestion
    default <T> CompletableFuture<T> aggregateAs(Predicate<? super HttpStatus> httpStatusFilter,
                                                 Function<String, T> jsonDecoder,
```

ikhoon reviewed (commented):
Could you migrate the existing methods that manually decoding a JSON content into an object such as: https://github.com/line/armeria/blob/6198d1f83a2a4b65a139e20c375f4149fad3d421/consul/src/main/java/com/linecorp/armeria/internal/consul/CatalogClient.java#L93-L99

karellen-kim committed:
Update core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

Update core/src/main/java/com/linecorp/armeria/common/HttpResponse.java

Co-authored-by: Ikhun Um <ih.pert@gmail.com>

karellen-kim commented on core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:
```diff
@@ -281,6 +284,21 @@ static HttpRequest of(RequestHeaders headers, Publisher<? extends HttpObject> pu
+     * @param path the path of the request
+     * @param content the content of the request
+     */
+    static HttpRequest ofJson(HttpMethod method, String path, Object content) {
```
Looks much better, thanks! @ikhoon

karellen-kim committed:
Provide a way to send a JSON object with WebClient (#3707)

karellen-kim committed:
Provide a way to send a JSON object with WebClient (#3707)

karellen-kim committed:
Provide a way to send a JSON object with WebClient (#3707)

ikhoon commented:
@karellen-kim I think #4021 takes over the features of this PR.
Let me close this PR. 🙇‍♂️ I hope to see the next PR in the near future. 😆

trustin commented:
But still we can't send a JSON object as it is, right? #4021 only takes care of the receiving part.

ikhoon commented:
Ah, this PR includes `HttpRequest.ofJson()` also. I missed that. Let me reopen and continue to develop the sending part.

ikhoon commented:
Gentle ping. @karellen-kim

ikhoon committed:
Merge branch 'master' into pr-3757@karellen-kim+json

jrhee17 commented:
It seems like this PR has been stale for a while, let me reassign @monorisk to finish this PR
